### ISSUE TITLE
Store panic payload in PanicException

### ISSUE THREAD

**Issue by davidhewitt** [PyO3/pyo3#853]
To improve ergonomics of `PanicException` introduced in #797, we would ideally restore the original panic payload with `panic::resume_unwind`.

This isn't trivial to do as far as I can tell - ideally I'd make `PanicException` a proper `#[pyclass]` and then we could store the `Box<dyn Any + Send + 'static>` payload inside it.

In the interest of getting #797 completed to unblock other work, I decided to open this as a separate issue for the future.

**Comment by kngwyu** [comment]
What kind of use cases do you have in your mind?
I'm not sure how to use stored panic.

**Comment by davidhewitt** [comment]
We should be calling `std::panic::resume_unwind` with the payload when we `fetch` a `PanicException` - see https://github.com/PyO3/pyo3/pull/797#discussion_r405343846

**Comment by davidhewitt** [comment]
Now that we've decided to wait to think on #797 a little more I'm going to close this issue; we can incorporate this in the final design of #797.
